### PR TITLE
[SelectionDAG][RISCV] Fix break of vnsrl pattern in issue #94265

### LINKED ISSUE #94265: [SDAG][RISCV] vnsrl pattern is broken after PR93182
See the diff of file `result/rvv/193fc4d8b7725d32.S` in https://github.com/dtcxzyw/llvm-codegen-benchmark/commit/c134d37047d3e6ee2336279936f98a6eab922e77#diff-bc8f186427aa6515f00cd4d764feea719d5d90240f5602f6a8acf2e114cc9660 

```
; llc -mtriple=riscv64 -mattr=+v test.ll -debug -o -
define <8 x i16> @func0000000000000000(<8 x i32> %0) #0 {
entry:
  %1 = sdiv <8 x i32> %0, <i32 64, i32 64, i32 64, i32 64, i32 64, i32 64, i32 64, i32 64>
  %2 = trunc <8 x i32> %1 to <8 x i16>
  %3 = shl <8 x i16> %2, <i16 10, i16 10, i16 10, i16 10, i16 10, i16 10, i16 10, i16 10>
  ret <8 x i16> %3
}
```
Before:
```
        vsetivli zero, 8, e32, m2, ta, ma
        vsra.vi v10, v8, 31
        vsrl.vi v10, v10, 26
        vadd.vv v8, v8, v10
        vsetvli zero, zero, e16, m1, ta, ma
        vnsrl.wi        v10, v8, 10
        vsll.vi v8, v10, 6
        ret
```
After #93182:
```
        vsetivli zero, 8, e32, m2, ta, ma
        vsra.vi v10, v8, 31
        vsrl.vi v10, v10, 26
        vadd.vv v8, v8, v10
        vsetvli zero, zero, e16, m1, ta, ma
        vnsrl.wi        v10, v8, 0
        vsll.vi v8, v10, 4
        li      a0, -1024
        vand.vx v8, v8, a0
        ret
```

I think the following code breaks vnsrl (i.e., trunc (srl X, C)) pattern:
https://github.com/llvm/llvm-project/blob/2ee7f49addb60f100929d10ad24ebf1743943b0d/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L2609-L2618

Do we need a TLI hook to query whether (trunc (srl X, C)) is natively supported?

cc @RKSimon @topperc

### PR DESCRIPTION
**Pull request by Fros1er**
Added a overload of `isTypeDesirableForOp` to take NewVT + OldVT, fixing the break of vnsrl described in issue #94265.

Fixes #94265

### CODE REVIEW

File: llvm/include/llvm/CodeGen/TargetLowering.h [hunk 1]
```diff
@@ -4339,6 +4339,20 @@ class TargetLowering : public TargetLoweringBase {
     return isTypeLegal(VT);
   }
 
+  /// Same as isTypeDesirableForOp(unsigned Opc, EVT VT), but also check if
+  /// the target is 'desirable' to truncate or extend OldVT to NewVT only using
+  /// the given node type, without the need of explicit trunc or ext. e.g. On
+  /// RISC-V Vector extension, vnsrl.wi can directly convert <n x i32> to <n x
+  /// i16> when shifting, with no extra trunc operations needed.
+  virtual bool isTypeDesirableForOp(unsigned Opc, EVT NewVT, EVT OldVT) const {
```
[the remark this answers is not in the document]
Comment (Fros1er) [reply llvm/include/llvm/CodeGen/TargetLowering.h:4347]: Maybe name like `isDesirableToCastOp` is better? As this function checks both promotion and truncation.

File: llvm/include/llvm/CodeGen/TargetLowering.h [hunk 2]
```diff
@@ -4339,6 +4339,21 @@ class TargetLowering : public TargetLoweringBase {
     return isTypeLegal(VT);
   }
 
+  /// Same as isTypeDesirableForOp(unsigned Opc, EVT VT), but also check if
+  /// the target is 'desirable' to truncate or extend OldVT to NewVT only using
+  /// the given node type, without the need of explicit trunc or ext. e.g. On
+  /// RISC-V Vector extension, vnsrl.wi can directly convert <n x i32> to <n x
+  /// i16> when shifting, with no extra trunc operations needed.
+  virtual bool isTypeDesirableForOpWithCast(unsigned Opc, EVT NewVT,
+                                            EVT OldVT) const {
```
Comment (RKSimon) [inline llvm/include/llvm/CodeGen/TargetLowering.h:4348]: Did you investigate whether we could just update the existing isTypeDesirableForOp? The x86 changes would be pretty easy and I suspect the AMDGPU changes would be as well

Comment (Fros1er) [reply llvm/include/llvm/CodeGen/TargetLowering.h:4348]: There are tons of usage of isTypeDesirableForOp, we need to make change everywhere if the parameter changes. Besides, as far as I know, it's rare for instructions to truncate operand while doing other jobs. Updating the existing one to NewVT + OldVT may be unnecessary for most of the cases.

Comment (RKSimon) [reply llvm/include/llvm/CodeGen/TargetLowering.h:4348]: Nearly all of those isTypeDesirableForOp calls are testing whether we should promote/truncate an existing op with the new type - including the old type as well is likely to be useful.

Comment (Fros1er) [reply llvm/include/llvm/CodeGen/TargetLowering.h]: I looked through the code and realized isTypeDesirableForOp is not suitable for the issue. isTruncateFree(SDValue, EVT) seems better here.

**Review by RKSimon — CHANGES_REQUESTED** [review]
This breaks a number of tests:
```
Failed Tests (8):
  LLVM :: CodeGen/AArch64/bitfield-extract.ll
  LLVM :: CodeGen/AArch64/extbinopload.ll
  LLVM :: CodeGen/AArch64/trunc-to-tbl.ll
  LLVM :: CodeGen/AArch64/zext-to-tbl.ll
  LLVM :: CodeGen/AMDGPU/amdgpu-codegenprepare-idiv.ll
  LLVM :: CodeGen/Hexagon/two-crash.ll
  LLVM :: CodeGen/Thumb2/shift_parts.ll
  LLVM :: CodeGen/X86/pr44915.ll
```

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -1884,6 +1884,16 @@ bool RISCVTargetLowering::isTruncateFree(EVT SrcVT, EVT DstVT) const {
   return (SrcBits == 64 && DestBits == 32);
 }
 
+bool RISCVTargetLowering::isTruncateFree(SDValue Val, EVT VT2) const {
+  // free truncate from vnsrl and vnsra
+  if (Subtarget.hasStdExtV() &&
+      (Val.getOpcode() == ISD::SRL || Val.getOpcode() == ISD::SRA) &&
+      Val.getValueType().isVector() && VT2.isVector()) {
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:1891]: Do we need to check that VT2 is exactly half the size of Val.getValueType()?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 4]
```diff
@@ -1884,6 +1884,19 @@ bool RISCVTargetLowering::isTruncateFree(EVT SrcVT, EVT DstVT) const {
...
 
+bool RISCVTargetLowering::isTruncateFree(SDValue Val, EVT VT2) const {
+  EVT SrcVT = Val.getValueType();
+  // free truncate from vnsrl and vnsra
+  if (Subtarget.hasStdExtV() &&
+      (Val.getOpcode() == ISD::SRL || Val.getOpcode() == ISD::SRA) &&
+      SrcVT.isVector() && VT2.isVector()) {
+    unsigned SrcBits = SrcVT.getVectorElementType().getSizeInBits();
+    unsigned DestBits = VT2.getVectorElementType().getSizeInBits();
+    return (SrcBits == DestBits * 2);
```
Comment (RKSimon) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:1895]: Are there any circumstances we'd want to drop down to TargetLowering::isTruncateFree(Val, VT2) if this is false?

**Review by RKSimon — APPROVED** [review]
LGTM with one minor query

**Review by RKSimon — APPROVED** [review]

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 5]
```diff
@@ -1884,6 +1884,21 @@ bool RISCVTargetLowering::isTruncateFree(EVT SrcVT, EVT DstVT) const {
   return (SrcBits == 64 && DestBits == 32);
 }
 
+bool RISCVTargetLowering::isTruncateFree(SDValue Val, EVT VT2) const {
+  EVT SrcVT = Val.getValueType();
+  // free truncate from vnsrl and vnsra
+  if (Subtarget.hasStdExtV() &&
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:1890]: I'm late to this, but shouldn't this use hasVInstructions(). It's excluding Zve*.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:1890]: Fixed by 3de76e4f57bdba612ddca7ac53b2f5787929391a

**Pull request merged**